flexmeasures/api/conftest.py

```diff
@@ -0,0 +1,39 @@
+import pytest
+
+from flask_login import login_user, logout_user
+
+from flexmeasures.api.tests.utils import UserContext
+
+
+@pytest.fixture
+def requesting_user(request):
+    """Use this fixture to log in a user for the scope of a test.
```

Comment:
1. Yes.
2. Yes, with the following exceptions:
    - `test_api_v3_0_users.test_logout` still uses the token instead of the new fixture; it tests logging out.
    - `test_assets_api.test_get_public_assets` still uses the token, together with the new fixture; the `public` endpoint had the `@auth_token_required` decorator. I added the following, which we can now discuss: `todo: consider changing this to @auth_required(), then simplify test_get_public_assets` (by removing the token from the test).
    - `test_sensor_schedules.test_trigger_and_get_schedule_with_unknown_prices` still uses the token, together with the new fixture; I believe the token is redundant here, and I'll try to remove it from the test.
    - `task_task_runs.test_api_task_run_post_unauthorized_wrong_role` still uses the token, together with the new fixture; I believe the token is redundant here, and I'll try to remove it from the test.

Reply:
Thanks for the details.

I feel we should test auth tokens at least once, also for logging in.

Reply:
Nope. I'm not opinionated. I'm just wondering whether the way we tested authentication before - where using a different token stood for authenticating a different user - was the way Flask-Login is supposed to work.

Reply:
I believe author tokens come with Flask-Security but I'm not sure.

In any case, it seems highly useful for API users of ours.

Reply:
[You are correct](https://flask-security-too.readthedocs.io/en/stable/features.html#token-authentication). I'm a bit lost on this topic now and would like to ask for your help.

Reply:
Well, I'm not sure which help I need to give.

So using auth tokens as the means to authenticate in the API is failing here?

That would mean you made tests work only by logging users in beforehand, so that the API knows them via some cookie/session mechanism.

I investigated the changes in flask-login for that version for a while, but wasn't able to find any change which might be responsible.

Is getting auth tokens to work what we need to do now?

Reply:
I think what broke when upgrading is that sending an auth token no longer authenticates you. Our permission checking decorators rely on checking the `current_user`, which just remains anonymous (unless you've actually send a login request previously).

At this point, I wouldn't know how to fix this. And even worse, I'm not even sure whether it is broken in the first place. It seems to me that we have been using token authentication as a way to manage session authentication on the client side. Is that conventional? I remember a possibly related discussion was put forth by @GustaafL about moving to long lived tokens.

Reply:
The discussion about long lived tokens is not related to this issue. The long lived token is in place of a user-password combination. So you would request the auth-token using the long lived token instead of with user-password. All subsequent calls to the API would still use the same authentication with the auth token.

I think the normal authentication for an API is to have an authentication flow (with either user-password or a refresh long lived token) that returns a short lived (30 minutes or less) token that can be used for the other available endpoints the API offers.

Reply:
Our token currently lives a few hours, I believe. We can configure it to live shorter.

These auth tokens are what our documentation describes - visit a login endpoint to get the token, then pass that on with your API requests. I don't know how the "session" way is working right now and if it is convenient for developers.

Reply:
I've spent half a day on this now (not my first time), not with a real solution. Maybe with a hack.

- When I use the FM client against an actual running FM instance (not running tests), authenticating via token seems to work. So this might be an issue with how the tests are set up (and more modern Flask ecosystem stumbles here, see [this](https://github.com/maxcountryman/flask-login/issues/704) for instance, where `app_context` is advised to be used differently than we have so far).
- In the example test by @Flix6x above, the user is correctly authenticated in flask_security, but to tell flask_login (0.6.2) about them fails.
- I can make the test (which uses auth token) pass with a hack in flask_security (explicitly calling ` flask_login.login_user(user)` in `decorators.py::_check_token`). Not sure, but we could [monkeypatch](https://docs.pytest.org/en/6.2.x/monkeypatch.html#simple-example-monkeypatching-functions) `_check_token` for testing. A bit icky, though ...

It's difficult to ask for help on this, as somehow many changes happened in the ecosystem of Flask (Flask / Werkzeug to begin with, but it includes Flask-Login and Flask-SQLAlchemy which apparently changed a lot. Flask-Login issues are closed when it's not about their internal code. Flask-Security seems not to have changed.

Thoughts:
- I can monkeypatch flask_security and we might be able to move on.
- I still wonder whether we will need `@auth_required` decorators more often. Either because ecosystem changes or because we didn't notice that they are needed as we luckily always had a session (and our tests might have spilled logins, like Felix seems to have seen). I'm not 100% about these things.
- Research into a better way to use `app_context` (or if that is even the problem) is appreciated. I'm burned out from this today, though.

Reply:
I'm interacting with the author/maintainer of Flask-Security [here](https://github.com/Flask-Middleware/flask-security/issues/834).

Reply:
@Flix6x I have added a few commits to this PR, which might be enough to (for now) move this PR forward. (I believe this discussion is the one holding it back).

They also include some explicit auth token checks we might have been missing so far, or which do have to be set explicitly in the newer versions, not 100% sure.
Some permission checks seem to have been missing in any case, good to have them now.

Please review and let me know.